### PR TITLE
feat(pypi): support GCloud credentials for Google Artifact Registry

### PR DESCRIPTION
**Pull request by maxbrunet**
## Changes

Make use of `getGoogleAuthToken` to populate `Authorization` header if registry is Google Artifact Registry (`.pkg.dev`)

## Context

* #30109

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/pypi/index.ts [hunk 1]
```diff
@@ -82,6 +86,21 @@ export class PypiDatasource extends Datasource {
     return dependency;
   }
 
+  private async getAuthHeaders(
+    lookupUrl: string,
+  ): Promise<OutgoingHttpHeaders> {
+    const url = new URL(lookupUrl);
```
Comment (rarkins) [inline lib/modules/datasource/pypi/index.ts:92]: Do we know 100% for sure that `lookupUrl` is validated as a URL by here? And if not then does `new URL()` throw, or return null?

Comment (maxbrunet) [reply lib/modules/datasource/pypi/index.ts]: Good catch, I wrapped it in a try-catch

**Review by rarkins — DISMISSED** [review]

**Review by secustor — COMMENTED** [review]
Any specific reason this is only scoped at Python packages?  

I guess `pkg.dev` has other registries outside of Python packages too.

File: lib/modules/datasource/pypi/__snapshots__/index.spec.ts.snap [hunk 2]
```diff
@@ -278,6 +278,49 @@ exports[`modules/datasource/pypi/index getReleases respects constraints 1`] = `
 }
 `;
 
+exports[`modules/datasource/pypi/index supports Google Auth with simple endpoint 1`] = `
```
Comment (viceice) [inline lib/modules/datasource/pypi/__snapshots__/index.spec.ts.snap:281]: revert this

File: lib/modules/datasource/pypi/index.ts [hunk 3]
```diff
@@ -4,14 +4,18 @@ import changelogFilenameRegex from 'changelog-filename-regex';
...
 import { regEx } from '../../../util/regex';
 import { ensureTrailingSlash } from '../../../util/url';
 import * as pep440 from '../../versioning/pep440';
 import { Datasource } from '../datasource';
 import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { getGoogleAuthToken } from '../util';
 import { isGitHubRepo, normalizePythonDepName } from './common';
 import type { PypiJSON, PypiJSONRelease, Releases } from './types';
 
+const googleArtifactRegistryDomain = '.pkg.dev';
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.ts:17]: inline

File: lib/modules/datasource/pypi/index.ts [hunk 4]
```diff
@@ -82,6 +86,27 @@ export class PypiDatasource extends Datasource {
     return dependency;
   }
 
+  private async getAuthHeaders(
+    lookupUrl: string,
+  ): Promise<OutgoingHttpHeaders> {
+    let url: URL;
+    try {
+      url = new URL(lookupUrl);
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.ts:94]: use `parseUrl` util function

File: lib/modules/datasource/pypi/index.spec.ts [hunk 5]
```diff
@@ -125,6 +130,64 @@ describe('modules/datasource/pypi/index', () => {
...
+        jest.fn().mockImplementationOnce(() => ({
+          getAccessToken: jest.fn().mockResolvedValue('some-token'),
+        })),
+      );
+      const res = await getPkgReleases({
+        ...config,
+        datasource,
+        packageName: 'azure-cli-monitor',
+      });
+      expect(res?.releases.pop()).toMatchObject({
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.spec.ts:157]:
```suggestion
      expect(res).toMatchObject({
```
compare
 full object

File: lib/modules/datasource/pypi/index.spec.ts [hunk 6]
```diff
@@ -125,6 +130,64 @@ describe('modules/datasource/pypi/index', () => {
...
+        jest.fn().mockImplementation(() => ({
+          getAccessToken: jest.fn().mockResolvedValue(undefined),
+        })),
+      );
+      const res = await getPkgReleases({
+        ...config,
+        datasource,
+        packageName: 'azure-cli-monitor',
+      });
+      expect(res?.releases.pop()).toMatchObject({
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.spec.ts:184]: same

File: lib/modules/datasource/pypi/index.spec.ts [hunk 7]
```diff
@@ -643,6 +706,44 @@ describe('modules/datasource/pypi/index', () => {
...
+      })),
+    );
+    expect(
+      await getPkgReleases({
+        datasource,
+        ...config,
+        constraints: { python: '2.7' },
+        packageName: 'dj-database-url',
+      }),
+    ).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.spec.ts:731]: no snapshots

File: lib/modules/datasource/pypi/index.spec.ts [hunk 8]
```diff
@@ -643,6 +706,44 @@ describe('modules/datasource/pypi/index', () => {
...
+  it('ignores an invalid URL when checking for auth headers', async () => {
+    const config = {
+      registryUrls: ['not-a-url/simple/'],
+    };
+    const res = await getPkgReleases({
+      ...config,
+      datasource,
+      packageName: 'azure-cli-monitor',
+    });
+    expect(res?.releases.pop()).toBeNil();
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.spec.ts:744]:
```suggestion
    expect(res).toBeNull();
```

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by maxbrunet** [comment]
> Any specific reason this is only scoped at Python packages?

Time and need

> I guess `pkg.dev` has other registries outside of Python packages too.

Yes[^1], Renovate already has support in the `docker` and `maven` datasources.

[^1]: https://cloud.google.com/artifact-registry/docs/supported-formats

**Review by viceice — APPROVED** [review]

**Comment by HonkingGoose** [comment]
This PR is accidentally breaking our `release` step on the `main` branch, see Discussion:

- https://github.com/renovatebot/renovate/discussions/31390

**Pull request merged**